src/GraphBLAS-sharp.Backend/Vector/COOVector/COOVector.fs

```diff
@@ -0,0 +1,542 @@
+
+                let gid = ndRange.GlobalID0
+
+                if gid < indicesArrayLength then
+                    let index = inputIndices.[gid]
+
+                    positions.[index] <- 0 @>
+
+        let kernel = clContext.Compile(preparePositions)
+
+        let creat = ClArray.create clContext workGroupSize
+
+        fun (processor: MailboxProcessor<_>) (inputIndices: ClArray<int>) (vectorSize: int) ->
+
+            let positions = creat processor vectorSize 1
+
+            let ndRange =
+                Range1D.CreateValid(inputIndices.Length, workGroupSize)
+
+            let kernel = kernel.GetKernel()
+
+            processor.Post(
+                Msg.MsgSetArguments(fun () -> kernel.KernelFunc ndRange inputIndices.Length inputIndices positions)
+            )
+
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+            positions
+
+    let setPositionsComplemented (clContext: ClContext) (workGroupSize: int) =
+
+        let setPositions =
+            <@ fun (ndRange: Range1D) length (positions: ClArray<int>) (resultIndices: ClArray<int>) ->
+
+                let gid = ndRange.GlobalID0
+
+                if gid = length - 1
+                   || gid < length
+                      && positions.[gid] <> positions.[gid + 1] then
+                    let index = positions.[gid]
+
+                    resultIndices.[index] <- gid @>
+
+        let kernel = clContext.Compile(setPositions)
+
+        let sum =
+            ClArray.prefixSumExcludeInplace clContext workGroupSize
+
+        let resultLength = Array.zeroCreate 1
+
+        fun (processor: MailboxProcessor<_>) (positions: ClArray<int>) ->
+
+            let prefixArrayLenght = positions.Length
+
+            let resultLengthGpu = clContext.CreateClCell 0
+
+            let _, r = sum processor positions resultLengthGpu
+
+            let resultLength =
+                let res =
+                    processor.PostAndReply(fun ch -> Msg.CreateToHostMsg<_>(r, resultLength, ch))
+
+                processor.Post(Msg.CreateFreeMsg<_>(r))
+
+                res.[0]
+
+            let resultIndices =
+                clContext.CreateClArray<int>(
+                    resultLength,
+                    hostAccessMode = HostAccessMode.NotAccessible,
+                    deviceAccessMode = DeviceAccessMode.ReadWrite,
+                    allocationMode = AllocationMode.Default
+                )
+
+            let ndRange =
+                Range1D.CreateValid(prefixArrayLenght, workGroupSize)
+
+            let kernel = kernel.GetKernel()
+
+            processor.Post(
+                Msg.MsgSetArguments(fun () -> kernel.KernelFunc ndRange prefixArrayLenght positions resultIndices)
+            )
+
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+            resultIndices
+
+    let complemented<'a when 'a: struct> (clContext: ClContext) (workGroupSize: int) =
```

Comment:
The vector is sparse, so implicit storage of complemented one requires a lot of memory. May be it's not the vector to decide whether a mask is complemented or not